File: server/services/chunker.js
import { v4 as uuidv4 } from 'uuid'

const CHUNK_SIZE   = 500  // words
const CHUNK_OVERLAP = 100 // words

/**
 * Split pages into overlapping word-based chunks.
 * @param {{ text, pageNumber, documentName }[]} pages
 * @returns {{ id, text, pageNumber, documentName }[]}
 */
export function chunkPages(pages) {
  const chunks = []

  for (const page of pages) {
    const words = page.text.split(/\s+/).filter(Boolean)
    if (words.length === 0) continue

    let start = 0
    while (start < words.length) {
      const end = Math.min(start + CHUNK_SIZE, words.length)
      const chunkText = words.slice(start, end).join(' ')

      chunks.push({
        id:           uuidv4(),
        text:         chunkText,
        pageNumber:   page.pageNumber,
        documentName: page.documentName,
      })

      if (end === words.length) break
      start += CHUNK_SIZE - CHUNK_OVERLAP
    }
  }

  return chunks
}

File: backend/services/chunker.py
import uuid
from typing import List, Dict


def chunk_text(
    pages: List[Dict],
    chunk_size: int = 500,
    overlap: int = 100,
) -> List[Dict]:
    """
    Split page-level text into overlapping word-based chunks.

    Each chunk carries:
      - chunk_id   : unique UUID
      - text       : the chunk content
      - page_number: source page
      - document_name: source file name
    """
    chunks: List[Dict] = []

    for page in pages:
        words = page["text"].split()

        if not words:
            continue

        if len(words) <= chunk_size:
            chunks.append({
                "chunk_id": str(uuid.uuid4()),
                "text": page["text"],
                "page_number": page["page_number"],
                "document_name": page["document_name"],
            })
        else:
            i = 0
            while i < len(words):
                window = words[i: i + chunk_size]
                chunks.append({
                    "chunk_id": str(uuid.uuid4()),
                    "text": " ".join(window),
                    "page_number": page["page_number"],
                    "document_name": page["document_name"],
                })
                if i + chunk_size >= len(words):
                    break
                i += chunk_size - overlap  # slide with overlap

    return chunks
